Generic dictionary minor performance improvement and simplification for R2R

fadimounir opened the pull request:
A set of refactoring changes to slighly improve the performance of generic dictionary access on R2R images, and simplifying the codebase:
    1) Removing dependency on CEEInfo::ComputeRuntimeLookupForSharedGenericTokenStatic (and deleting the API).
    2) Stop parsing the generic type/method signatures when generating the R2R dictionary lookup assembly stub.
    3) Avoid re-encoding the generic type/method signatures in a new in-memory blob using a SigBuilder  (signatures used directly from the R2R image)
    4) Moved the parsing/loading of type/method signatures to Dictionary::PopulateEntry()
    5) Dictionary index and slot number are now encoded in the generated assembly stub instead of the signature

@jkotas, @JohnChen0  PTAL.

@rahku, we'll need to fix the arm64 assembly stub to pass the 3rd argument to the new JIT helper APIs.

fadimounir commented:
@dotnet-bot test Windows_NT Checked pri1r2r please

fadimounir commented:
@dotnet-bot test Ubuntu Checked pri1r2r please

fadimounir commented:
@dotnet-bot test Ubuntu arm64 please

jkotas commented on src/inc/corinfo.h:
```diff
@@ -614,10 +614,12 @@ enum CorInfoHelpFunc
+    CORINFO_HELP_RUNTIMEHANDLE_METHOD_LOG,      // determine a type/field/method handle at run-time, with IBC logging
+    CORINFO_HELP_RUNTIMEHANDLE_METHOD_WITHSLOT, // determine a type/field/method handle at run-time, with dictionary slot
+    CORINFO_HELP_RUNTIMEHANDLE_CLASS,           // determine a type/field/method handle at run-time
+    CORINFO_HELP_RUNTIMEHANDLE_CLASS_LOG,       // determine a type/field/method handle at run-time, with IBC logging
```
These helpers should not be in this enum because of the JIT does not care about them. There are other similar helpers like this, e.g. `JIT_StaticFieldAddress_Dynamic`.

jkotas commented on src/vm/genericdict.cpp:
```diff
@@ -575,22 +631,78 @@ Dictionary::PopulateEntry(
     *ppSlot = NULL;
 
+    Module * pContainingZapModule = ExecutionManager::FindZapModule(dac_cast<TADDR>(signature));
+    Module * pReadyToRunModule = ExecutionManager::FindReadyToRunModule(dac_cast<TADDR>(signature));
```
There tend to be expensive operations, especially for large workloads with a lot of modules (it is a liner search). Can you avoid them by passing the module around as necessary instead?

jkotas commented:
> slighly improve the performance 

It would useful to do some validation that there is actual improvement (e.g. make up microbenchmark with a lot of generic types and drill into it with perfview if necessary).

kyulee1 commented:
Does Arm64 code (CreateDictionaryLookupHelper) need to be implemented as well? Or is the third argument optional?
/cc @rahku 
@dotnet-bot test Windows_NT arm64 checked

fadimounir commented:
@kyulee1 The arm64 implementation will need to be updated as well. @rahku is already looking into it.

kyulee1 commented:
@fadimounir Got it. I missed the description.

fadimounir commented:
I did some validation on my end, and I without the fix the speed is about 9% slower.

---
In reply to: [225268535](https://github.com/dotnet/coreclr/pull/5690#issuecomment-225268535) [](ancestors = 225268535)

fadimounir commented:
@kyulee1, @rahku, I have pushed a new iteration of the PR which does not require any codegen changes for the arm64 stub.

fadimounir commented:
@dotnet-bot test Windows_NT arm64 checked

fadimounir commented:
@dotnet-bot test Ubuntu Checked pri1r2r please

fadimounir commented:
@dotnet-bot test Windows_NT Checked pri1r2r please

jkotas commented:
>  without the fix the speed is about 9% slower.

What was the benchmark that you have measured?

kyulee1 commented:
@dotnet-bot test Windows_NT arm64 Checked

fadimounir commented:
@jkotas I wrote a small benchmark test that does ldtoken(T), ldtoken(List<T>) from a dictionary slot, and another ldtoken instruction which will never have a slot in the dictionary. In order to measure the performance of the new code vs. old code, I removed the patching of the R2R cell so that we always call DynamicHelperWorker(), and removed the patching of the dictionary cell so that it's always NULL, so we always call the JIT helper to load the generic type handle (did that on 2 builds, one with the fix, and the other without)

fadimounir commented:
@dotnet-bot test OSX x64 Checked Build and Test please

fadimounir commented:
@dotnet-bot test Ubuntu x64 Checked R2R pri1 Build & Test please

fadimounir commented:
@dotnet-bot test Ubuntu x64 Checked R2R pri1 Build & Test please

fadimounir commented:
@dotnet-bot test Ubuntu Checked pri1r2r please
